### ISSUE TITLE
Realize clause listing/[0,1] in Prolog itself

### ISSUE THREAD

**Issue by jburse** [jburse/jekejeke-devel#216]
We have a couple of new requirements for listing which
probably become increasingly difficult when an attempt
is made to solve them in Java.

There are also new insights about the dynamic database
access. It turns out that clause/2 wouldn't be suitable to
realize listing for the followng reasons:

- **Static Access:**
   clause/2 only allows access to dynamic, thread_local
   and group_local predicates. So it cannot be used for listing/[0,1]
   that mainly shows static predicates. So we need to bypass this
   restriction.

- **var/1 Indexing:**
   It turns out that clause/2, even when we would allow static predicates,
   does filter var/1 indexing. So a call to clause(p(X1,..,Xn),Y) would 
   not allow retrieving all clauses. This is about to change for dynamic, 
   thread_local and group_local predicates, but will remain for static 
   predicates. So we need to bypass this restriction.

- **Callable Property:**
  Listing needs also to filter the sys_context/1 callable property.
  If we issue clause(p(X1,..,Xn),Y) we will not retrieve the genuine
  functor of each clause and can therefore not determine the
  geniune callable property of each clause. So we need to bypass 
  this restriction.

A solution to all these fundamental access problems could
be a new predicate rule/2. But it would not take Head and Body, but
rather Indicator and Clause:
```
% rule(+Indicator, -Clause)
```
That SWI-Prolog walks similar paths with its new rule/[2,3] is 
rather accidential. Our rule/2 would allow access to the method
listClauses() which is already used in the Java implementation of 

listing/[0,1]. So this is not some new functionality. The only novel
thing is interfacing it into Prolog itself. And evident measure needed
to realize listing/[0,1] in Prolog itself.

**Comment by jburse** [comment]
The following changes might profit from this change:

**Use rebuild framework to unwrap tabling #188**
[https://\github.com/jburse/jekejeke-devel/issues/188](https://github.com/jburse/jekejeke-devel/issues/188)

**Comment by jburse** [comment]
We already have a working prototype of a new predicate sys_list/3.
But we cannot list specials or foreign predicates with it, and we also
have no meta information to do it via Prolog. So one idea is to

generate a directive (:- special(...)). (:- foreign), etc... The new listing
predicate wouldn't be aware that built-ins are handled like this
and the development listing, that list the automatic generated 

could work the same.

**Comment by jburse** [comment]
There is some progress in realizing this change. We also moved the
predicate properties listing to SpecialPred. We should do the same
for the operator listing and move it to SpecialOper.

To complete the change we need to have library(friendly) also rewritten,
only then can we replace the bulk of native Java code around
SpecialLoad.showClause().

When completed we should do regression testing of library(automatic)
and library(friendly). Also documentation has to be adapted since
we shortened documentation of listing/0 and listing/1 into one

paragraph, same for library(automatic).
